src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+use spirv::*;
+
+pub(crate) struct Module {
+    pub(crate) physical_layout: PhysicalLayout,
+    pub(crate) logical_layout: LogicalLayout,
+}
+
+impl Module {
+    pub(crate) fn new(module: &crate::Module) -> Self {
+        let version: Word = (0x0u32 << 24)
+            | ((module.header.version.0 as u32) << 16)
+            | ((module.header.version.1 as u32) << 8)
+            | module.header.version.2 as u32;
+
+        Module {
+            physical_layout: PhysicalLayout::new(version, module.header.generator, 0x0u32),
+            logical_layout: LogicalLayout::new(),
+        }
+    }
+}
+
+pub(crate) struct PhysicalLayout {
+    magic_number: Word,
+    version: Word,
+    generator: Word,
+    pub(crate) bound: Word,
+    instruction_schema: Word,
+}
+
+impl PhysicalLayout {
+    pub(crate) fn new(version: Word, generator: Word, instruction_schema: Word) -> Self {
+        PhysicalLayout {
+            magic_number: MAGIC_NUMBER,
+            version,
+            generator,
+            bound: 0,
+            instruction_schema,
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<Word> {
+        let mut words: Vec<Word> = vec![];
+
+        words.push(self.magic_number);
+        words.push(self.version);
+        words.push(self.generator);
+
+        match self.bound {
+            0 => panic!("Bound cannot be 0"),
```

Comment:
Spec: 
```
0 < id < Bound

Bound should be small, smaller is better, with all <id> in a module being densely packed and near 0.
```
By this I concluded that a Bound never is allowed to be zero. I haven't found this explicitly stated in the spec, so I could be wrong.

Reply:
What happens if an empty module is saved?
I think we should just write 1 instead of panicking

Reply:
That would go against the SPIR-V spec in my opinion. As with `1` the program is saying that it does contain an instruction with a result id. 

Are there any cases where an user passes in an empty module? If so, maybe we can abort the parsing silently, and hint in the console, something like that?

Reply:
No, that's not contradicting the spec. The bound is literally a higher bound. The spec even says "bound *should* be small". It is valid for it to be anything as long as all the IDs are lower.

Reply:
Okay, so I think I get what you mean. Bound is always: [total_id] + 1. So if we don't have any ids (aka instructions), that means that bound = 0 + 1 = 1, right? So not having instructions results in a Bound of 1.

Reply:
That's not what I'm saying:)
Bound is at least as high as the highest used ID, +1.
So if you have IDs 0, 1, 3, then bound can be 4, or it can be 10. Both are valid choices.
If you have no IDs, 0 is not a valid choice by the spec, but 1 is, still. I.e. 1 is still higher than "any used IDs" :)

Reply:
However, in the Validation rules of SPIR-V it says: "There is at least one OpEntryPoint instruction, unless the Linkage capability is being used". I'm not familiar with "Linkage", but for the rest of the capabilities, it is actually necessary to have at least one instruction. 

That being said, I actually think we don't need to check here for the Bound, but we need to check if there is one Entry Point available.

For the time being, I think the best we can do is just returning `1` when the match matches `0`.

Reply:
I gave this some more though. and the "0" match is even unreachable, as the bound is now always set at `self.id_count + 1`. I will remove the match.

Reply:
yeah, that makes sense :+1: